File: packages/merkle-tree/src/test/sizeAddition.test.ts
import * as crypto from 'crypto';
import { 
  addItem, 
  createTree,
  HashedItem
} from '../lib/merkle-tree';
import { visualizeTree } from '../lib/visualize';

describe('Size calculation with file addition', () => {
  
  // Helper function to create a file hash with specific size
  function createHashedItem(name: string, content: string, size: number): HashedItem {
    const hash = crypto.createHash('sha256')
      .update(content)
      .digest();
    return {
      name,
      hash,
      length: size,
      lastModified: new Date(),
    };
  }

  test('leaf node should have size equal to file length', () => {
    // Create a tree with one file
    const fileSize = 1024;
    const fileHash = createHashedItem('test.txt', 'test content', fileSize);
    const tree = addItem(createTree("12345678-1234-5678-9abc-123456789abc"), fileHash);

    // Verify the leaf node has correct size
    expect(tree.sort?.size).toBe(fileSize);
    expect(tree.sort?.size).toBe(fileSize);
  });

  test('parent node should have size equal to sum of children', () => {
    // Create a tree with two files of different sizes
    const file1Size = 500;
    const file2Size = 1000;
    
    let tree = createTree("12345678-1234-5678-9abc-123456789abc");
    tree = addItem(tree, createHashedItem('file1.txt', 'content 1', file1Size));
    tree = addItem(tree, createHashedItem('file2.txt', 'content 2', file2Size));
    
    // Verify root node has sum of all file sizes
    expect(tree.sort?.size).toBe(file1Size + file2Size);
    
    // Verify left child has correct size
    expect(tree.sort?.left?.size).toBe(file1Size);
    
    // Verify right child has correct size
    expect(tree.sort?.right?.size).toBe(file2Size);
    
    // Verify metadata reflects total size
    expect(tree.sort?.size).toBe(file1Size + file2Size);
  });

  test('sizes are propagated correctly in a multi-level tree', () => {
    // Create a tree with multiple files
    let tree = createTree("12345678-1234-5678-9abc-123456789abc");
    const sizes = [100, 200, 300, 400, 500, 600, 700];
    const totalSize = sizes.reduce((sum, size) => sum + size, 0);
    
    // Add files with different sizes
    for (let i = 0; i < sizes.length; i++) {
      tree = addItem(tree, createHashedItem(`file${i}.txt`, `content ${i}`, sizes[i]));
    }
    
    expect(tree.sort?.size).toBe(totalSize);    
    expect(tree.sort?.size).toBe(totalSize);    
    expect(tree.sort?.left?.size).toBe(1500);
    expect(tree.sort?.right?.size).toBe(1300);
  });
  
  test('size is updated when adding files', () => {
    let tree = createTree("12345678-1234-5678-9abc-123456789abc");
    expect(tree.sort).toBeUndefined();
    
    // Add first file
    const file1Size = 1000;
    tree = addItem(tree, createHashedItem('file1.txt', 'content 1', file1Size));
    expect(tree.sort?.size).toBe(file1Size);
    
    // Add second file
    const file2Size = 2000;
    tree = addItem(tree, createHashedItem('file2.txt', 'content 2', file2Size));
    expect(tree.sort?.size).toBe(file1Size + file2Size);
    
    // Add third file
    const file3Size = 3000;
    tree = addItem(tree, createHashedItem('file3.txt', 'content 3', file3Size));
    expect(tree.sort?.size).toBe(file1Size + file2Size + file3Size);
  });
});

File: packages-zig/merkle-tree-zig/src/test/sizeAddition.test.zig
//
// Size calculation with file addition (port of src/test/sizeAddition.test.ts).
//

const std = @import("std");
const merkle_tree_zig = @import("merkle-tree-zig");
const merkle_verify = @import("merkle-verify.zig");
const merkle_tree = merkle_tree_zig.merkle_tree;

test "leaf node should have size equal to file length" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const fileSize: u64 = 1024;
    const emptyTree = merkle_tree.createTree(merkle_verify.TEST_TREE_ID);
    const tree = try merkle_tree.addItem(allocator, &emptyTree, try merkle_verify.createSha256HashedItem(allocator, "test.txt", "test content", fileSize));
    try std.testing.expectEqual(fileSize, tree.sort.?.size);
}

test "parent node should have size equal to sum of children" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = merkle_tree.createTree(merkle_verify.TEST_TREE_ID);
    tree = try merkle_tree.addItem(allocator, &tree, try merkle_verify.createSha256HashedItem(allocator, "file1.txt", "content 1", 500));
    tree = try merkle_tree.addItem(allocator, &tree, try merkle_verify.createSha256HashedItem(allocator, "file2.txt", "content 2", 1000));

    try std.testing.expectEqual(@as(u64, 1500), tree.sort.?.size);
    try std.testing.expectEqual(@as(u64, 500), tree.sort.?.left.?.size);
    try std.testing.expectEqual(@as(u64, 1000), tree.sort.?.right.?.size);
}

test "sizes are propagated correctly in a multi-level tree" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = merkle_tree.createTree(merkle_verify.TEST_TREE_ID);
    const sizes = [_]u64{ 100, 200, 300, 400, 500, 600, 700 };
    for (sizes, 0..) |size, index| {
        const name = try std.fmt.allocPrint(allocator, "file{d}.txt", .{index});
        const content = try std.fmt.allocPrint(allocator, "content {d}", .{index});
        tree = try merkle_tree.addItem(allocator, &tree, try merkle_verify.createSha256HashedItem(allocator, name, content, size));
    }

    try std.testing.expectEqual(@as(u64, 2800), tree.sort.?.size);
    try std.testing.expectEqual(@as(u64, 1500), tree.sort.?.left.?.size);
    try std.testing.expectEqual(@as(u64, 1300), tree.sort.?.right.?.size);
}

test "size is updated when adding files" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = merkle_tree.createTree(merkle_verify.TEST_TREE_ID);
    try std.testing.expect(tree.sort == null);

    tree = try merkle_tree.addItem(allocator, &tree, try merkle_verify.createSha256HashedItem(allocator, "file1.txt", "content 1", 1000));
    try std.testing.expectEqual(@as(u64, 1000), tree.sort.?.size);

    tree = try merkle_tree.addItem(allocator, &tree, try merkle_verify.createSha256HashedItem(allocator, "file2.txt", "content 2", 2000));
    try std.testing.expectEqual(@as(u64, 3000), tree.sort.?.size);

    tree = try merkle_tree.addItem(allocator, &tree, try merkle_verify.createSha256HashedItem(allocator, "file3.txt", "content 3", 3000));
    try std.testing.expectEqual(@as(u64, 6000), tree.sort.?.size);
}
